### PR TITLE
Ft python approximation curve

### PR DESCRIPTION
**Pull request by jzwar**
# Overview
Small Change that enables approximations of splines to a point cloud. The code reuses as much of the existing code as possible and introduces only little modifications to the original structure.  Implementation based on NURBS-book. Allows for arbitrary B-Splines and returns Residual

# Testing
Implementation has been tested with respect to python code from [NURBS-Python](https://github.com/orbingol/NURBS-Python)

### CODE REVIEW

File: Sources/Fittings/interpolate.cpp [hunk 1]
```diff
@@ -64,11 +75,95 @@ void InterpolateSurface(double* points,
...
+  }
+
+  // Solve linear system with LU-decomposition
+  control_points = LUSolve(LSmatrix, NtR, n_ctps_v, dim);
+  
+  // Add P0 at front and Pm at end (might be inefficient)
+  for (int i{0}; i < dim ; i++){
+    control_points.insert(control_points.begin(), points[dim-i-1]);
+    control_points.push_back(points[(num_points-1) * dim + i]);
+  }
```
Comment (jzwar) [inline Sources/Fittings/interpolate.cpp:148]:
```suggestion
  // Add P0 at front and Pm at end (might be inefficient)
  control_points.resize(control_points.size() + dim * 2);
  for (int i{n_ctps_v * dim - 1}; i >= 0; i--) {
    control_points[i+dim] = control_points[i];
  }
  for (int i{0}; i < dim ; i++) {
    control_points[i]  = points[i];
    control_points[(n_ctps_v +1) * dim + i] = points[(num_points-1) * dim + i];
  }
```

This is actually more efficient, my bad...

**Review by j042 — CHANGES_REQUESTED** [review]
Thank you! Looking forward to approximate curves.

File: Sources/Fittings/utils.hpp [hunk 2]
```diff
@@ -129,12 +143,12 @@ inline std::vector<double> ComputeKnotVector(int& degree,
...
                               std::vector<double>& knot_vector,
-                              int& num_control_points,
-                              double& knot) {
+                              const int& num_control_points,
+                              const double& knot) {
 
     int span = degree + 1;
     while (span < num_control_points && knot_vector[span] <= knot) {
-        span += 1;
+        ++span;
```
[the remark this answers is not in the document]
Comment (jzwar) [reply Sources/Fittings/utils.hpp:151]: In this particular case there is no difference. I just prefer this over the other stylewise :D

File: Sources/Fittings/utils.hpp [hunk 3]
```diff
@@ -101,26 +101,40 @@ inline void ParametrizeSurface(double* points,
         v_l[v] /= size_u;
     }
 
-    delete[] pts_u, pts_v;
+    delete pts_u, pts_v;
```
[the remark this answers is not in the document]
Comment (jzwar) [reply Sources/Fittings/utils.hpp]: According to [this](https://stackoverflow.com/questions/1484739/how-to-properly-delete-a-pointer-to-array) you seem to be correct. I will change this

File: Sources/Fittings/fitting.cpp [hunk 4]
```diff
@@ -0,0 +1,176 @@
+#include "Sources/Fittings/fitting.hpp"
+
+double FitCurve(double* points,
+                      int& num_points,
```
Comment (j042) [inline Sources/Fittings/fitting.cpp:4]: alignment

**Review by j042 — CHANGES_REQUESTED** [review]
1.0

**Pull request merged**